WIP- Merge Throttle into Develop

jaypacamarra opened the pull request:
Improved and added to the throttle task according to issue #8.
I created the Throttle driver that does these things:

- BSE valid range check
- APPS 1 and APPS 2 valid range check
- Brake Plausibility Check
- 10% APPS difference check
- Driver_Controls driver that packages all checks into nice functions (i.e. getPedalReadings(), check10Percent...)
- Everything stored in the VCU data structure

Any feedback is welcome. Thanks.

Commit:
Added additional BSE and APPS fault flag variables

- Added additional fault flags in the VCU data structure
- Initialized said fault flag variables in the vcu_data.c to a value of zero

Commit:
Added short checks for APPS1 and APPS2

- Changed BSE_MIN_VALUE
- Added APPS1 and APPS2 MIN and MAX define values
- Added Check for short to GND/5V on APPS sensor 1

Commit:
Fixed bug: Not entering tasks

- Changed config queue length to 4 (freeRTOS config header)
- Changed timer task stack depth to 256 (freeRTOS config header)
- Enabled pin 0, pin 1, pin 2 in ADC1 Group 1

Commit:
Added APPS1 and APPS2 10% difference fault

- Changed FP_sensor_diff to type float from type uint
- I will test this

Commit:
Added test comment

Commit:
Changed 10DIFF fault comment

- Added tags folder to .gitignore file for use with VIM and ctags

Commit:
Add brake pedal plausibility checking

Checks if APPS1 and APPS2 agree within 10%

Commit:
Unnermge develop and add Brake pedal plausibility checking

Commit:
Add 100ms fault timer for APPS1 and APPS2 10DIFF

- Flickering between fault and no fault still occurs

Commit:
Remove flickering of PC_UART fault message

- Flickering of UART messages removed if not toggling the GIO user LEDs at the same time

Commit:
Enable GIOA[6] to test brakelight

Commit:
Fix logic of brake plausiblity check

- Calculation is now based on APPS1 and APPS2 defined min and max values

Commit:
Change the fault counter variable more meaningful

Commit:
Perecent accel pedal pressed is 0 when sensor is below min value

Commit:
Add fault timer for BSE, APPS1, APPS2 range

Commit:
Add APPS/BSE plausibility reset if less than 5% pedal

Commit:
Add lowpass signal conditioning

The alpha value is still a placeholder value

Commit:
Clarify short circuit check comments

Commit:
Add hysteresis to remove brake light flickering near threshhold

- Needs to be tested

Commit:
Remove else statement because causes flickering

- Tested, no flickering noticed
- Flickering may occur if power supply connection the BSE ADC input pin is loose

Commit:
Add 10% APPS Implausibility for O-0 and O-1

- Higher optimization levels (O-2 and above breaks the APPS 10% DIFF fault check)

Commit:
Software timer on APPS1 range fault check

Commit:
Upgrade timer implementation in throttle task

- Increased NUMBER_OF_TIMERS from 2 to 6 in main.c
- news timers are xTimers[2],xTimers[3],xTimers[4],xTimers[5]
- Removed timer implementation using xFreqeuncy

Commit:
Add check if xTimers[2-5] is created

Commit:
Replace xTimerReset with xTimerStop

- xTimerReset restarts the timer which doesn't make sense for the case where a fault is no longer occuring so xTimerStop stops the timer

Commit:
Change comments

Commit:
Fix FAULT TIMER not returning false when fault clears

Commit:
debugging

Commit:
update includes

Commit:
Merge branch 'develop' into Throttle_Task-JayP

Commit:
Add #if macro for LV_MONITOR or else gets stuck in lv_monitorInit()

Commit:
Increase Heap Size so don't freeze when mutexes are enabled

Commit:
Setup Mibspi for DAC testing

Commit:
DAC works for testing

Commit:
Redefine brake light port and clean up code a bit for brake implausibility

Commit:
Add Throttle driver and clean up throttle task

Commit:
Change THROTTLE driver name to Driver_controls according to issue#8

Commit:
Add driver function to calculate pedal press percentages

- Change throttle calculation

Commit:
Add comments on percentages in vcu data structure

Commit:
Add execution_timer.h

Commit:
Add EEPROM Driver

Commit:
Revert "Add EEPROM Driver"

This reverts commit 8d0b838b58fe9cf07ea764a33f4c66f2eb302bb0.

Commit:
Revert "Add execution_timer.h"

This reverts commit ff487861392d7f9606bd4c67b3d04121417ab3a8.

Commit:
Fix big where BSE range fault flag wasnt getting updated

Commit:
Add execution_timer.h again

Commit:
Fix stuff

Commit:
Add throttle padding/buffering for eliminating unintended range faults

Commit:
Add vcuKey mutex

Commit:
Change board define

Commit:
Increase padding from 5% to 8%

Commit:
Adds padding and define more macros in throttle header

Defined the following in throttle.h:
- ALPHA
- macros for the padded MIN and MAX
- HYSTERESIS

Commit:
Merge branch 'develop' of https://github.com/sfuphantom/vcu-fw into develop

Commit:
Merge branch 'develop' into Throttle_Task-JayP

Commit:
Merge develop into throttle task

Commit:
Change driver name to Throttle

Junaidk11 commented on VCU/Phantom/Drivers/RTOS/phantom_freertos.h:
```diff
@@ -40,6 +40,10 @@
 /* This timer is used to debounce the interrupts for the RTDS and SDC signals */
 void Timer_300ms(TimerHandle_t xTimers);
 void Timer_2s(TimerHandle_t xTimers);
+void APPS1_SEVERE_RANGE_FAULT_CALLBACK(TimerHandle_t xTimers);      // prototype for APPS1 severe range fault software timer callback - jaypacamarra
+void APPS2_SEVERE_RANGE_FAULT_CALLBACK(TimerHandle_t xTimers);      // prototype for APPS2 severe range fault software timer callback - jaypacamarra
+void BSE_SEVERE_RANGE_FAULT_CALLBACK(TimerHandle_t xTimers);        // prototype for BSE severe range fault software timer callback - jaypacamarra
+void FP_DIFF_SEVERE_FAULT_CALLBACK(TimerHandle_t xTimers);          // prototype for FP difference by 10% fault software timer callback - jaypacamarra
```
You have 6 timer callbacks, change NUMBER_OF_TIMERS at line 38 to 6.

jaypacamarra replied:
Oops. Fixed that.

Junaidk11 commented on VCU/Phantom/tasks/source/task_throttle.c:
```diff
@@ -75,164 +70,136 @@ extern volatile uint8_t initializationOccured;
+        {
+            // update throttle percentage in vcu data structure
+            apps_percent_avg = (Percent_APPS1_Pressed + Percent_APPS2_Pressed) / 2;
+            VCUDataPtr->AnalogOut.throttle_percentage.value = apps_percent_avg;
```
You returned the vcuKey before this statement. So, you should retake the vcuKey before you make any changes to the VCU Data structure.

jaypacamarra replied:
Alright I made a commit that retakes the vcu key back there and then returns the vcu key right afterward.

Junaidk11 reviewed (commented):
Your task looks good. Just change the NUMBER_OF_TIMERS to 6 because I saw 6 timer callbacks. Also, you probably forgot to retake the vcuKey after you returned it and tried to make changes to the VCU structure.

Guojiaxi commented on VCU/Phantom/tasks/source/task_throttle.c:
```diff
@@ -75,164 +70,136 @@ extern volatile uint8_t initializationOccured;
+    APPS1_range_fault_timer_started = false;
+    APPS2_range_fault_timer_started = false;
+    BSE_range_fault_timer_started   = false;
+    FP_diff_fault_timer_started     = false;
```
These statements do not need to be here as they are already initialized to false during their declarations.

jaypacamarra replied:
okay I removed them.

Guojiaxi commented on VCU/Phantom/tasks/source/task_throttle.c:
```diff
@@ -5,65 +5,60 @@
+extern bool APPS1_range_fault_timer_started = false;
+extern bool APPS2_range_fault_timer_started = false;
+extern bool BSE_range_fault_timer_started   = false;
+extern bool FP_diff_fault_timer_started     = false;
```
These declarations should only be moved to throttle.c without the externs. The only time these variables are being referred in this file is from lines 82-86, but those lines are redundant as these variables are already initialized to begin with.

Guojiaxi replied:
Also, the only file that is concerned with these timer flags anyways is throttle.c so why not have them contained in throttle.c?

jaypacamarra replied:
Alright I fixed that, I removed those 4 and just put them in throttle.c without the externs. I'm not sure why I did that.

Guojiaxi commented on VCU/Phantom/tasks/source/task_throttle.c:
```diff
@@ -5,65 +5,60 @@
+// To store percent pressed of foot pedals
+extern float Percent_APPS1_Pressed;
+extern float Percent_APPS2_Pressed;
+extern float Percent_BSE_Pressed;
```
All these variables that are being extern'd from throttle.c can be received from getters in throttle.c instead, as they are only being read in this file. Also, it is bad practice to be accessing things from another module/file without #including them.

jaypacamarra replied:
Alright nice catch. I fixed this.

Guojiaxi commented on VCU/Phantom/Drivers/Throttle/throttle.c:
```diff
@@ -0,0 +1,405 @@
+extern bool APPS1_range_fault_timer_started;
+extern bool APPS2_range_fault_timer_started;
+extern bool BSE_range_fault_timer_started;
+extern bool FP_diff_fault_timer_started;
```
These timer flags don't need to be extern'd as stated in comment regarding the flags in task_throttle.c

jaypacamarra replied:
Yup changed that. I removed the extern now.

Guojiaxi commented on VCU/Phantom/Drivers/Throttle/throttle.c:
```diff
@@ -0,0 +1,405 @@
+
+float volatile Percent_APPS1_Pressed;
+float volatile Percent_APPS2_Pressed;
+float volatile Percent_BSE_Pressed;
```
These variables (sensor sums and the press percents) should have getter functions so other modules/files can read these values safely.

something like this could work:
```
unsigned int Throttle_getBSESensorSum()
{
    return BSE_sensor_sum;
}
```

jaypacamarra replied:
Alright I added those getters.

Guojiaxi reviewed (commented):
The code seems like it will work without these changes. My comments are just suggestions to make throttle.c a little more modular and code cleaner. Otherwise, looks good to me.

Commit:
Change NUMBER_OF_TIMERS from 2 to 6

Commit:
Retake vcu key when updating throttle % in vcu data structure

vcu key is given back after updating the throttle %

Commit:
Remove timer started variables init to false

Also, I now included my throttle driver. I didn't notice I didn't even use it.

Commit:
Remove fault timer started variables

I also removed apps_percent_avg and throttle as global variables and made them local to the "Set Throttle" portion at the bottom starting in line 173.

Commit:
Create getters for apps1, apps2, and bse pedal percent

I used them in task_throttle.c now.

Commit:
Define getters for the sensor sums

Commit:
Use BSE sensor sum getter in phantom_freertos.c

Guojiaxi commented on VCU/Phantom/tasks/source/task_throttle.c:
```diff
@@ -36,29 +37,13 @@ extern unsigned int volatile BSE_sensor_sum;
 extern unsigned int volatile FP_sensor_1_sum;
 extern unsigned int volatile FP_sensor_2_sum;
```
task_throttle.c doesn't need references to BSE_sensor_sum, FP_sensor_1_sum, and FP_sensor_2_sum anymore, because the getters for them exists in throttle.h.

Guojiaxi replied:
Also, you might want change the references to BSE_sensor_sum on line 114 and 126 to what the getter returns.

jaypacamarra replied:
Done. Thanks for the feedback.

Guojiaxi reviewed (commented):
Other than these final (I think) changes, the task is turning out to look pretty clean!

Commit:
set up mibspi1 for throttle on DAC

Commit:
Use het1 pin 12 for brakelight

Commit:
eeprom driver compile error

Commit:
Fix compile error. throttle to THROTTLE

Commit:
Test brakelight works on vcu, dac still not tested

Commit:
VCU DAC works

- DAC output from 0.6V - 4.4V depending on APPS

Commit:
Remove extern to bse, fp1 and fp2 sums in task_throttle.c

- Also change reference to BSE_sensor_sum with Throttle_getBSESensorSum() in line 111 and line 123

Guojiaxi reviewed (approved):
lgtm

jaypacamarra commented:
Thanks Josh for the approval! I will quickly make sure everything is still good, don't wanna mess up develop! Then I'll merge pull request.

Commit:
Cleaned up code before merge into develop
- Commented out debugging code
- included "Throttle.h" in phantom_freertos.c
